### DISCUSSION TITLE
cannot pickle 'Decimal' object when using Pythonnet

### DISCUSSION THREAD

**Discussion by Omnistic** [zauberzeug/nicegui#2985] [Q&A] [answered]
### Question

Hello everyone,

A few days ago I posted [my question](https://stackoverflow.com/questions/78389008/cannot-pickle-decimal-object-when-using-pythonnet-and-nicegui) to StackOverflow without success (so far). I thought I would also ask the community here (and update my StackOverflow accordingly).

I am using Pythonnet to interface some hardware (as per the manufacturer recommendation) and I wanted to write a small user interface with NiceGUI. I also wanted my button to disable while my hardware initializes. I started with the [disable button with a context manager example](https://nicegui.io/documentation/button#disable_button_with_a_context_manager). But faced an issue with the `run.cpu_bound` method, which returns a `TypeError: cannot pickle 'Decimal' object` when I do the following:

```py
from contextlib import contextmanager
from nicegui import run, ui
import clr
from System import Decimal

@contextmanager
def disable(button: ui.button):
    button.disable()
    try:
        yield
    finally:
        button.enable()

class my_class:
    def __init__(self, value):
        self.value = Decimal(value)

def some_function():
    return my_class(1)

async def get_slow_response(button: ui.button) -> None:
    with disable(button):
        return_value = await run.cpu_bound(some_function)

ui.button('Get slow response', on_click=lambda e: get_slow_response(e.sender))

ui.run(port=80)
```

I need the `System.Decimal` to send commands to my hardware but if I remove it, the UI works normally. Is there perhaps a way to disable my button that doesn't involve `run.cpu_bound`? Any help is appreciated :)

**Comment by falkoschindler** [comment]
Hi @Omnistic,

As you might already found out, `run.cpu_bound` needs to pickle all data for sending it to another process and receiving the result. Somehow `System.Decimal` can't be pickled. Actually I'm not sure where `System` is coming from. Is `from decimal import Decimal` an option?

In general, I'd recommend calling `run.cpu_bound` with simple functions that work with "normal" datatypes that can be pickled.

**Comment by Omnistic** [comment]
Hi @falkoschindler,

I think this `System` is part of `pythonnet` somehow. The strange thing is that I have to `import clr` before importing `System` otherwise it doesn't work. I am not so advanced in programming I must say. The hardware I am interfacing is using C# libraries that have been compiled to DLLs and the manufacturer is using `pythonnet` to send the data to the hardware, hence the `Decimal` (?). Is there a way to keep my function with `Decimal` and still get the behavior I want: the button is disabled during the function execution? I'd really like to use `nicegui` because it is super easy to understand and has a very neat appearance :)

Thanks for your help.

**Comment by Omnistic** [comment, marked as answer]
I don't really understand why (I'm not a programmer), but using `asyncio.to_thread` instead of `run.cpu_bound` seems to work better. I couldn't use the context manager somehow but I'm happy with this solution.
```python
import asyncio
import clr
import time
from nicegui import ui
from System import Decimal

class my_class:
    def __init__(self, value):
        self.value = Decimal(value)

def some_function():
    time.sleep(1)
    return my_class(1)

async def get_slow_response(button: ui.button) -> None:
    button.disable()
    return_value = await asyncio.to_thread(some_function)
    button.enable()

ui.button('Get slow response', on_click=lambda e: get_slow_response(e.sender))

ui.run(port=80)
```

**Comment by rodja** [comment]
Beware: Python has GIL, which means cpu intensive tasks on a thread will block all execution of the process. You'll need to spawn a new Process to offload cpu intensive tasks. This is what `run.cpu_bound` does and therefore requires pickling the parameters and return values.

**Reply by Omnistic** [reply]
@rodja 

Thanks a lot. Would you mind explaining this to me in Layman's terms? In my applications, `my_class` is associated with a hardware. In this instance, it is a motorized linear stage. In the constructor of the class, I "home" the stage, that means a command is sent via serial communication to the stage controller and when the stage controller has finished "homing" the stage, it returns a serial command as well. I don't think the task is computationally intensive because the controller takes care of it (but it easily takes 4-5 seconds for the stage to move to its limits and find the "home" position).

I am not at all a Python developer, but I use Python for scientific computations and sometimes creating those user interfaces. I am aware that `nicegui` was probably not designed with that application in mind, but its simplicity was what drew my attention. I don't intend to become a Python developer, but if you have some resources to share that might help me be more familiar with those terms, please let me know :)

**Reply by falkoschindler** [reply]
Hi @Omnistic, using multi-_threading_ instead of multi-_processing_ seems correct in your case, since you're just waiting for an actuator to move instead of doing heavy computation. So you could use `run.io_bound` for this purpose. It doesn't move the function to another process and, thus, doesn't need to pickle its inputs and outputs. For CPU-bound functions it's necessary to use `run.cpu_bound` in order to make use of additional CPU cores.

For more information, have a look at our [FAQ](https://github.com/zauberzeug/nicegui/wiki/FAQs#why-is-my-long-running-function-blocking-ui-updates) about this topic and the NiceGUI documentation https://nicegui.io/documentation/section_action_events#running_cpu-bound_tasks and https://nicegui.io/documentation/section_action_events#running_i_o-bound_tasks.

**Reply by Omnistic** [reply]
@falkoschindler, thank you for sharing those resources. What's the difference between `asyncio.to_thread` and `run.io_bound`? I am a bit reluctant to change my code since it works now.

**Reply by falkoschindler** [reply]
I'm not quite sure. `run.cpu_bound` is based on asyncio's `run_in_executor`, but I don't know `to_thread`. Feel free to keep using `to_thread` if it works for you. 👍

**Reply by Omnistic** [reply]
@falkoschindler it works also with `run.io_bound` :)
